cou/apps/auxiliary.py

```diff
@@ -288,15 +289,15 @@ class RabbitMQServer(AuxiliaryApplication):
 
     def pre_upgrade_steps(
         self, target: OpenStackRelease, units: Optional[list[Unit]]
-    ) -> list[PreUpgradeStep]:
+    ) -> list[UpgradeStep]:
```

Comment:
In the end of the day `PreUpgradeStep`, `UnitUpgradeStep`, `PostUpgradeStep` are the `UpgradeStep`, but I don't see a reason why to change the type here and in other parts of the code. Can't we keep the way that it is right now?

Reply:
I had to, because mypy was not happy.  Ideally we would remove  PreUpgradeStep, UnitUpgradeStep, and PostUpgradeStep and replace them all with `UpgradeStep`; the distinction is not useful.

Reply:
As far as I know I agree that it's not useful. It was a bet in the past that it could possibly help us and I guess in the end it didn't help at all.

With that said, I don't feel comfortable changing the type in those functions just because it makes more convenient for this PR and let the code into an inconsistency state where some places still uses it. I know that refactoring that will increase substantially the number of lines changed and unit tests will break.

IMHO I think reusing the implementaion of `_get_wait_step` is a wrong approach because of the following reasons:

- That was designed especially for `PostUpgradeStep` and some application like ceph-osd, ocatvia waits for the whole model to be active/idle. We don't need that behavior for just refreshing the charm. Just checking the app itself would be enough in this case
- The time can be way larger than necessary (not a real issue), but it feels awkward to see that we are going to wait like 2400 seconds for a simple charm refresh.

I think we can simply manually add the wait step and keep the current types. E.g:

```python
    def _get_refresh_charm_steps(self, target: OpenStackRelease) -> list[PreUpgradeStep]:
        """Get steps for refreshing the charm.
        :param target: OpenStack release as target to upgrade
        :type target: OpenStackRelease
        :raises ApplicationError: When application has unexpected channel.
        :return: Steps for refreshing the charm
        :rtype: list[PreUpgradeStep]
        """
        wait_step = PreUpgradeStep(
            description="Wait up to 120s for app '{self.name}' to reach idle state",
            parallel=False,
            coro=self.model.wait_for_idle(self.wait_timeout, apps=[self.name])
         )
        if self.is_from_charm_store:
            return [self._get_charmhub_migration_step(target), wait_step]
        ....
```

Reply:
> I don't feel comfortable changing the type in those functions just because it makes more convenient for this PR and let the code into an inconsistency state where some places still uses it. 

It's quite the opposite in this PR - I've changed the type of a function, and updated types across the codebase accordingly so it is consistent.

> I think reusing the implementaion of _get_wait_step is a wrong approach

I see your point here, that makes sense.  I was hoping to reduce duplication by reusing the wait logic, but if the wait logic can be simpler for these cases, I'll update it there too.

Reply:
Updated, thanks for your comments :)

Reply:
> It's quite the opposite in this PR - I've changed the type of a function, and updated types across the codebase accordingly so it is consistent.

You would still have `PreUpgradeStep` on other parts of the project like hypervisors, plan and etc. That was the  inconsistency that I was talking about.

Closing this thread once the approach changed.